libdevcore/CommonData.h

```diff
@@ -50,17 +50,46 @@ enum class HexPrefix
 	DontAdd = 0,
 	Add = 1,
 };
+
+enum class HexCase
+{
+	Lower = 0,
+	Upper = 1,
+	Mixed = 2,
+};
+
 /// Convert a series of bytes to the corresponding string of hex duplets.
 /// @param _w specifies the width of the first of the elements. Defaults to two - enough to represent a byte.
 /// @example toHex("A\x69") == "4169"
 template <class T>
-std::string toHex(T const& _data, int _w = 2, HexPrefix _prefix = HexPrefix::DontAdd)
+std::string toHex(T const& _data, int _w = 2, HexPrefix _prefix = HexPrefix::DontAdd, HexCase _case = HexCase::Lower)
 {
 	std::ostringstream ret;
-	unsigned ii = 0;
-	for (auto i: _data)
-		ret << std::hex << std::setfill('0') << std::setw(ii++ ? 2 : _w) << (int)(typename std::make_unsigned<decltype(i)>::type)i;
-	return (_prefix == HexPrefix::Add) ? "0x" + ret.str() : ret.str();
+	unsigned n = 0;
+	for (auto i: _data) {
+		// switch hex case every four hexchars
+		auto c = std::nouppercase;
+		if (_case == HexCase::Upper) c = std::uppercase;
+		else if (_case == HexCase::Mixed) c = ((( n++ & 2) == 0 ) ? std::uppercase : std::nouppercase);
+		ret << std::hex << c
+			<< std::setfill('0') << std::setw(n++ ? 2 : _w) << (int)(typename std::make_unsigned<decltype(i)>::type)i;
```

Comment:
```suggestion
			<< std::setfill('0') << std::setw(_w) << (int)(typename std::make_unsigned<decltype(i)>::type)i;
```
I think we don't need this test and increment. Also, this double increment leads to the mixed case being every two digits not four. If you remove this (as suggested in this comment) and remove the loop below (as suggested in the other comment) it should work.